common/firestore/FirestoreProtoClient.java

```diff
@@ -0,0 +1,270 @@
+      e.printStackTrace();
+    }
+  }
+
+  public FirestoreProtoClient(String project, String token) {
+    FirebaseOptions options =
+        new FirebaseOptions.Builder()
+            .setCredentials(GoogleCredentials.of(new AccessToken(token, null)))
+            .setProjectId(project)
+            .build();
+    FirebaseApp.initializeApp(options);
+    client = FirestoreClient.getFirestore();
+  }
+
+  public Firestore getClient() {
+    return client;
+  }
+
+  private String joinPath(String collection, String documentId) {
+    if (collection.endsWith("/")) {
+      return collection + documentId;
+    }
+    return collection + "/" + documentId;
+  }
+
+  private Message parseProto(DocumentSnapshot document, Message.Builder builder)
+      throws InvalidProtocolBufferException {
+    return builder
+        .build()
+        .getParserForType()
+        .parseFrom(Base64.getDecoder().decode(document.getString(PROTO_FIELD)));
+  }
+
+  private CollectionReference getCollectionReference(String[] parts, int length) {
+    DocumentReference docRef;
+    CollectionReference collectionRef = client.collection(parts[0]);
+    for (int i = 1; i < length; i += 2) {
+      docRef = collectionRef.document(parts[i]);
+      collectionRef = docRef.collection(parts[i + 1]);
+    }
+    return collectionRef;
+  }
+
+  public DocumentReference getDocumentReference(String path) {
+    String[] parts = path.split("/");
+    if (parts.length % 2 != 0) {
+      throw new IllegalArgumentException("path length should be even but is " + parts.length);
+    }
+    return getCollectionReference(parts, parts.length - 1).document(parts[parts.length - 1]);
+  }
+
+  public DocumentReference getDocumentReference(String collection, String documentId) {
+    return getDocumentReference(joinPath(collection, documentId));
+  }
+
+  public ApiFuture<DocumentSnapshot> getDocumentAsync(String path) {
+    return getDocumentReference(path).get();
+  }
+
+  public ApiFuture<DocumentSnapshot> getDocumentAsync(String collection, String documentId) {
+    return getDocumentAsync(joinPath(collection, documentId));
+  }
+
+  public DocumentSnapshot getDocument(String path) {
+    try {
+      return getDocumentAsync(path).get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  public DocumentSnapshot getDocument(String collection, String documentId) {
+    return getDocument(joinPath(collection, documentId));
+  }
+
+  public Message getProtoDocument(String path, Message.Builder builder) {
+    try {
+      return parseProto(getDocument(path), builder);
+    } catch (IOException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  public Message getProtoDocument(String collection, String documentId, Message.Builder builder) {
+    return getProtoDocument(joinPath(collection, documentId), builder);
+  }
+
+  public ApiFuture<WriteResult> setDocumentAsync(String path, Map map) {
```

Comment:
Why the map is not parameterized? Don't we know even the type of key?

Reply:
I think it can potentially be a string or a map, but I limited it to a string for now, since I think it's what we'll be using.